s3torchconnector/src/s3torchconnector/s3iterable_dataset.py

```diff
@@ -133,6 +148,26 @@ def _get_transformed_object(self, bucket_key: S3BucketKeyData) -> Any:
         )
 
     def __iter__(self) -> Iterator[Any]:
-        return map(
-            self._get_transformed_object, self._get_dataset_objects(self._get_client())
+        worker_id = 0
+        num_workers = 1
+        if not self._share_dataset_within_process:
+            worker_info = torch.utils.data.get_worker_info()
+            if worker_info is not None:
+                worker_id = worker_info.id
+                num_workers = worker_info.num_workers
+
+        self._shard_index = num_workers * self._rank + worker_id
+        self._shard_count = num_workers * self._world_size
+
+        if self._shard_index == 0 and self._shard_count == 1:
+            return map(
+                self._get_transformed_object,
+                self._get_dataset_objects(self._get_client()),
+            )
+
+        sharded_objects = (
```

Comment:
I think pickling happens before DataLoader calls `__iter__` for the first time, otherwise all the workers will access the same `worker_info` from parent worker.